### PR TITLE
Support handling the error status of SFP modules

### PR DESCRIPTION
**Pull request by stephenxs**
1. Update get_change_event
   - Return event in a bitmap format in case there is an error
   - Return an extra map indexed by sfp_error in case there is a Mellanox-specific error
   - Indicate whether the error is a blocking error
2. Expose the error status of SFP modules to CLI
   - A platform API calls SDK API to fetch the error status
     - Display the SFP module state: plugged, unplugged and plugged with error.
     - The SFP error state is displayed only if the SFP module state is plugged with error
   - The CLI will call platform API and provide user-friendly output
3. Redirect stderr to /dev/null in order to eliminate errors in case SFP module is not plugged
4. Beautify the code, making all the state/error/descriptions defined in a uniform place

Signed-off-by: Stephen Sun <stephens@nvidia.com>

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

### CODE REVIEW

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+def mock_get_sdk_handle(self):
+    if not self.sdk_handle:
+        self.sdk_handle = 1
+    return self.sdk_handle
+
+def mock_get_sfp_error_code(self):
+    return self.oper_code, self.error_code
+
+
+device_info.get_platform = mock_get_platform
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/tests/test_sfp.py:35]: device_info.get_platform = MagicMock(return_value='x86_64-mlnx_msn2410-r0')

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+SFP._read_eeprom_specific_bytes = mock_read_eeprom_specific_bytes
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/tests/test_sfp.py:36]: SFP._read_eeprom_specific_bytes = MagicMock(return_value=None)

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+SFP._get_error_code = mock_get_sfp_error_code
+Chassis.get_sdk_handle = mock_get_sdk_handle
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/tests/test_sfp.py:38]: Chassis.get_sdk_handle = MagicMock(return_value=1)

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 4]
```diff
@@ -388,7 +406,7 @@ def get_presence(self):
     # Read out any bytes from any offset
     def _read_eeprom_specific_bytes(self, offset, num_bytes):
         eeprom_raw = []
-        ethtool_cmd = "ethtool -m sfp{} hex on offset {} length {}".format(self.index, offset, num_bytes)
+        ethtool_cmd = "ethtool -m sfp{} hex on offset {} length {} 2>/dev/null".format(self.index, offset, num_bytes)
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:409]: why do this change?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 5]
```diff
@@ -2158,3 +2176,66 @@ def is_replaceable(self):
...
+                1: cls.SFP_MLNX_ERROR_DESCRIPTION_LONGRANGE_NON_MLNX_CABLE,
+                2: cls.SFP_ERROR_DESCRIPTION_I2C_STUCK,
+                3: cls.SFP_ERROR_DESCRIPTION_BAD_EEPROM,
+                4: cls.SFP_MLNX_ERROR_DESCRIPTION_ENFORCE_PART_NUMBER_LIST,
+                5: cls.SFP_ERROR_DESCRIPTION_UNSUPPORTED_CABLE,
+                6: cls.SFP_ERROR_DESCRIPTION_HIGH_TEMP,
+                7: cls.SFP_ERROR_DESCRIPTION_BAD_CABLE,
+                8: cls.SFP_MLNX_ERROR_DESCRIPTION_PMD_TYPE_NOT_ENABLED,
+                12: cls.SFP_MLNX_ERROR_DESCRIPTION_PCIE_POWER_SLOT_EXCEEDED,
+                255: cls.SFP_STATUS_OK
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2213]: Why we need this?

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 6]
```diff
@@ -2158,3 +2176,66 @@ def is_replaceable(self):
...
+        oper_status, error_code = self._get_error_code()
+        if oper_status == SX_PORT_MODULE_STATUS_INITIALIZING:
+            error_description = self.SFP_STATUS_INITIALIZING
+        elif oper_status == SX_PORT_MODULE_STATUS_PLUGGED:
+            error_description = self.SFP_STATUS_OK
+        elif oper_status == SX_PORT_MODULE_STATUS_UNPLUGGED:
+            error_description = self.SFP_STATUS_UNPLUGGED
+        elif oper_status == SX_PORT_MODULE_STATUS_PLUGGED_DISABLED:
+            error_description = self.SFP_STATUS_DISABLED
+        elif oper_status == SX_PORT_MODULE_STATUS_PLUGGED_WITH_ERROR:
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2235]: else ? Or do we need a "else" branch?

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+def test_sfp_partial_and_then_full_initialize():
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/tests/test_sfp.py:40]: Is this test case related to the feature?

Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/tests/test_sfp.py:40]: No. They are codes that already in the master but not in your branch. You might want to rebase your branch.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py (hunk 5 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2213]: Do you mean `cls.SFP_STATUS_OK`? 255 is also a value that can be returned from `PMAOS` register. It's defined in PRM.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py (hunk 4 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:409]: Redirecting the `stderr` to `/dev/null` avoids errors printing to the CLI. Those errors occur when an SFP module doesn't exist and can corrupt the output of `show error status`.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 8]
```diff
@@ -17,6 +17,7 @@ class MockSxFd(object):
     new_sx_fd_t_p = MagicMock(return_value=MockSxFd())
     new_sx_user_channel_t_p = MagicMock()
 from sonic_py_common.logger import Logger
+from sonic_platform.sfp import SFP
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: Fixed

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 9]
```diff
@@ -69,19 +60,38 @@ class MockSxFd(object):
...
+]
+
+SDK_ERRORS_TO_ERROR_BITS = {
+    0x0: SFP.SFP_ERROR_BIT_POWER_BUDGET_EXCEEDED,
+    0x2: SFP.SFP_ERROR_BIT_I2C_STUCK,
+    0x3: SFP.SFP_ERROR_BIT_BAD_EEPROM,
+    0x5: SFP.SFP_ERROR_BIT_UNSUPPORTED_CABLE,
+    0x6: SFP.SFP_ERROR_BIT_HIGH_TEMP,
+    0x7: SFP.SFP_ERROR_BIT_BAD_CABLE,
+    0x1: 0x00010000,
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: Replaced with constant definition.

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py (hunk 1 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/tests/test_sfp.py:35]: This is not part of this PR. Will open another PR to address the comment.

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py (hunk 2 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/tests/test_sfp.py:36]: This is not part of this PR. Will open another PR to address the comment.

File: platform/mellanox/mlnx-platform-api/tests/test_sfp.py (hunk 3 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/tests/test_sfp.py:38]: This is not part of this PR. Will open another PR to address the comment.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py (hunk 6 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2240]: Added an `else` branch.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py [hunk 10]
```diff
@@ -2158,3 +2183,68 @@ def is_replaceable(self):
...
+            error_code: The error code returned by _get_error_code
+
+        Returns:
+            The error description
+        """
+        oper_status, error_code = self._get_error_code()
+        if oper_status == SX_PORT_MODULE_STATUS_INITIALIZING:
+            error_description = self.SFP_STATUS_INITIALIZING
+        elif oper_status == SX_PORT_MODULE_STATUS_PLUGGED:
+            error_description = self.SFP_STATUS_OK
```
Comment (Junchao-Mellanox) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2237]: cls.SFP_MLNX_ERROR_DESCRIPTION_RESERVED ?

Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py:2237]: No. This is the case the SFP module is plugged without any error, which should be “OK”.
The above one is the case SFP module is plugged with errors but the error code is 255 which means reserved.

**Review by Junchao-Mellanox — APPROVED** [review]
